Support statistics dump summary only mode

kusmour opened the pull request:
Added a new --summary option to statistics dump command so that it is much light weight than the full version.
Also modified the `SymbolFileDWARFDwo *GetDwoSymbolFile` as suggested in [PR #80218](https://github.com/llvm/llvm-project/pull/80218#discussion_r1473639878)

kusmour committed:
Support statistics dump summary only mode

Summary:
Added a new --summary option to statistics dump command so that it is much light weight than the full version.
With this change, statistics dump --summary can now be included in lldb command line telemetry without slowing down lldb exiting.

hawkinsw commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -100,60 +101,91 @@ llvm::json::Value ConstStringStats::ToJSON() const {
+
+    json::Array breakpoints_array;
+    double totalBreakpointResolveTime = 0.0;
+    // Rport both the normal breakpoint list and the internal breakpoint list.
```
```suggestion
    // Report both the normal breakpoint list and the internal breakpoint list.
```

hawkinsw commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -74,7 +75,7 @@ json::Value ModuleStats::ToJSON() const {
   if (!symfile_modules.empty()) {
     json::Array symfile_ids;
-    for (const auto symfile_id: symfile_modules)
+    for (const auto symfile_id : symfile_modules)
```
Is this just a whitespace change?

kusmour replied:
Yes I can revert that, probs a formatting on save that I didn't notice.

hawkinsw commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.h:
```diff
@@ -186,6 +186,10 @@ class SymbolFileDWARF : public SymbolFileCommon {
   GetMangledNamesForFunction(const std::string &scope_qualified_name,
                              std::vector<ConstString> &mangled_names) override;
 
+  // Return total currently loaded debug info
```
```suggestion
  // Return total currently loaded debug info.
```

hawkinsw commented on lldb/source/Commands/Options.td:
```diff
@@ -1412,4 +1412,7 @@ let Command = "trace schema" in {
     Desc<"Include statistics for all targets.">;
+  def statistics_dump_summary: Option<"summary", "s">, Group<1>,
+    Desc<"Dump only high level summary statistics."
+         "Exclude targets, modules, breakpoints etc.. details.">;
```
```suggestion
         "Exclude targets, modules, breakpoints etc... details.">;
```

hawkinsw commented on lldb/source/Commands/Options.td:
```diff
@@ -1412,4 +1412,7 @@ let Command = "trace schema" in {
   def statistics_dump_all: Option<"all-targets", "a">, Group<1>,
     Desc<"Include statistics for all targets.">;
+  def statistics_dump_summary: Option<"summary", "s">, Group<1>,
+    Desc<"Dump only high level summary statistics."
```
```suggestion
    Desc<"Dump only high-level summary statistics."
```

hawkinsw reviewed (commented):
Thank you for doing this! I hope that these little nits are helpful!

kusmour committed:
address nit comments

clayborg commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -86,9 +86,13 @@ class LLDB_API SBTarget {
   /// \return
   ///     A SBStructuredData with the statistics collected.
-  lldb::SBStructuredData GetStatistics();
+  lldb::SBStructuredData GetStatistics(bool summary_only = false);
```
This violates our public API policy where we can't take away or change a previously available public API function. So we must:
- keep the old function around
- add a new function with extra parameters

If we are going to possibly add more options to GetStatistics in the future, we don't want to have to keep overloading this function with new values since we would need to keep adding overloads. The way we fix this in the API is to create a SBStatisticsOptions class that has accessors. This way we can keep adding more features in the future as it is ok to add new methods to any existing classes. So this code should look like:
```
  /// Returns a dump of the collected statistics.
  ///
  /// \return
  ///     A SBStructuredData with the statistics collected.
  lldb::SBStructuredData GetStatistics();

  /// Returns a dump of the collected statistics.
  ///
  /// \param[in] options
  ///   An objects object that contains all options for the statistics dumping.
  ///
  /// \return
  ///     A SBStructuredData with the statistics collected.
  lldb::SBStructuredData GetStatistics(SBStatisticsOptions options);
```
Then we need to create a simple public API class for `SBStatisticsOptions`:
```
class SBStatisticsOptions {
public:
  SBStatisticsOptions();
  bool GetSummaryOnly();
  void SetSummaryOnly(bool b);
};
```
And if later we might want to add another bool option, we can just add the new methods to this class. Lets say later we want to only emit target stats for a specific target, instead of all targets, we could add new functions to `SBStatisticsOptions`:
```
class SBStatisticsOptions {
public:
  SBStatisticsOptions();
  bool GetSummaryOnly();
  void SetSummaryOnly(bool b);
  /// Limit statistics data for only 1 target instead of all targets
  void SetTarget(SBTarget target);
};
```
And we can still use our `lldb::SBStructuredData GetStatistics(SBStatisticsOptions options);` function because the signature didn't change!

clayborg commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -326,7 +330,7 @@ class LLDB_API SBTarget {
   uint32_t GetAddressByteSize();
 
   const char *GetTriple();
-  
+
```
revert whitespace only changes

clayborg commented on lldb/include/lldb/Target/Statistics.h:
```diff
@@ -133,7 +133,7 @@ struct ConstStringStats {
 class TargetStats {
 public:
-  llvm::json::Value ToJSON(Target &target);
+  llvm::json::Value ToJSON(Target &target, bool summary_only = false);
```
We probably want a lldb_private::StatisticsOptions class that contains all options:
```
struct StatisticsOptions {
  bool summary_only = false;
};
```
Then when you create the SBStatisticsOptions class, it should contain a `std::unique_ptr<lldb_private::StatisticsOptions> m_opaque_up;` member, just like all other public API classes. And the methods on `SBStatisticsOptions` will modify the contained `lldb_private::StatisticsOptions` object appropriately. The reason we need the `SBStatisticsOptions` class to have a single member that is a unique pointer is so if we add extra things to `lldb_private::StatisticsOptions`, it won't change the size of the `SBStatisticsOptions` class since it just has a `unique_ptr<lldb_private::StatisticsOptions>` as its only member variable, which makes it suitable for use in the public API. We can't change the size of our objects if we want to maintain a stable public API.

clayborg commented on lldb/include/lldb/Target/Statistics.h:
```diff
@@ -171,9 +171,14 @@ class DebuggerStats {
   /// \return
   ///     Returns a JSON value that contains all target metrics.
-  static llvm::json::Value ReportStatistics(Debugger &debugger, Target *target);
+  static llvm::json::Value ReportStatistics(Debugger &debugger, Target *target,
+                                            bool summary_only = false);
```
pass in const `lldb_private::StatisticsOptions &options` instead of bool

clayborg commented on lldb/include/lldb/Target/Target.h:
```diff
@@ -1599,7 +1599,7 @@ class Target : public std::enable_shared_from_this<Target>,
   /// \return
   ///     Returns a JSON value that contains all target metrics.
-  llvm::json::Value ReportStatistics();
+  llvm::json::Value ReportStatistics(bool summary_only = false);
```
pass in `const lldb_private::StatisticsOptions &options` instead of bool

clayborg commented on lldb/source/API/SBTarget.cpp:
```diff
@@ -197,17 +197,18 @@ SBDebugger SBTarget::GetDebugger() const {
 }
 
-SBStructuredData SBTarget::GetStatistics() {
+SBStructuredData SBTarget::GetStatistics(bool summary_only) {
```
pass in const `SBStatisticsOptions options` instead of bool

clayborg commented on lldb/source/Commands/CommandObjectStats.cpp:
```diff
@@ -83,13 +86,15 @@ class CommandObjectStatsDump : public CommandObjectParsed {
     }
 
     bool m_all_targets = false;
+    bool m_summary_only = false;
```
Change this to:
```
lldb_private::StatisticsOptions m_options;
```
We should probably also move `m_all_targets` into `lldb_private::StatisticsOptions`.

kusmour replied:
> We should probably also move `m_all_targets` into `lldb_private::StatisticsOptions`

I can do that but notice the `m_all_targets` was not used anywhere outside of the command object. The target is decided with the current execution context.

bulbazord commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -86,9 +86,13 @@ class LLDB_API SBTarget {
+  ///
   /// \return
   ///     A SBStructuredData with the statistics collected.
-  lldb::SBStructuredData GetStatistics();
```
This is an ABI-breaking change. You'll want to add a new method entirely.

clayborg commented on lldb/source/Commands/CommandObjectStats.cpp:
```diff
@@ -83,13 +86,15 @@ class CommandObjectStatsDump : public CommandObjectParsed {
       return llvm::ArrayRef(g_statistics_dump_options);
     }
 
     bool m_all_targets = false;
```
This should get moved into `lldb_private::StatisticsOptions` as an option. Add accessor to `SBStatisticsOptions` so we can get/set this as well.

bulbazord commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFUnit.h:
```diff
@@ -241,7 +241,7 @@ class DWARFUnit : public UserID {
   FileSpec::Style GetPathStyle();
 
-  SymbolFileDWARFDwo *GetDwoSymbolFile();
+  SymbolFileDWARFDwo *GetDwoSymbolFile(bool load_if_needed = true);
```
Instead of having a bool here, you could make this an enum with 2 values. Something like this:
```
enum LoadMode : bool { eDoNotForceLoad, eForceLoad };
```

bulbazord commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.h:
```diff
@@ -186,6 +186,10 @@ class SymbolFileDWARF : public SymbolFileCommon {
+  // Return total currently loaded debug info.
+  // For cases like .dwo files, the debug info = skeleton debug info + all dwo
+  // debug info where .dwo files might not be loaded yet. Calling this function
+  // will not force the loading of any .dwo files.
```
Could you make this into a proper doxygen comment? With `///` instead of `//`

jeffreytan81 commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -2687,7 +2687,7 @@ uint64_t SymbolFileDWARF::GetDebugInfoSize() {
       continue;
 
-    SymbolFileDWARFDwo *dwo = cu->GetDwoSymbolFile();
+    SymbolFileDWARFDwo *dwo = cu->GetDwoSymbolFile(false);
```
Instead of hard-coding `false` here, I think it should the caller of `SymbolFileDWARF::GetDebugInfoSize` caller (statistics.cpp in this case) making this decision.

kusmour replied:
Will put a new PR to address this

clayborg commented on lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:
```diff
@@ -2687,7 +2687,7 @@ uint64_t SymbolFileDWARF::GetDebugInfoSize() {
       continue;
 
-    SymbolFileDWARFDwo *dwo = cu->GetDwoSymbolFile();
+    SymbolFileDWARFDwo *dwo = cu->GetDwoSymbolFile(false);
```
This bool might need to be a bool in the "SymbolFile::GetDebugInfoSize()" virtual function. Users might want the know the rigtht answer (exactly how much total debug info they have if `load_if_needed == true` or just what is loaded currently `load_if_needed == false`

kusmour replied:
In that case I will leave this commit out of this PR and start a separate one :)

clayborg commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -100,60 +101,91 @@ llvm::json::Value ConstStringStats::ToJSON() const {
 
-json::Value TargetStats::ToJSON(Target &target) {
-  CollectStats(target);
+json::Value TargetStats::ToJSON(Target &target, bool summary_only) {
```
Pass in "const lldb_private::StatisticsOptions &options" here insetead of bool

clayborg replied:
This function seems to still take a "bool summary_only"?

clayborg commented on lldb/source/Target/Target.cpp:
```diff
@@ -4962,4 +4962,6 @@ std::recursive_mutex &Target::GetAPIMutex() {
 
 /// Get metrics associated with this target in JSON format.
-llvm::json::Value Target::ReportStatistics() { return m_stats.ToJSON(*this); }
+llvm::json::Value Target::ReportStatistics(bool summary_only) {
```
pass in "const lldb_private::StatisticsOptions &options" here instead of bool

bulbazord reviewed (changes requested):
Haven't looked over everything yet but this has an ABI-breaking change. Please revert the existing changes to SBTarget.

kusmour committed:
Create SBStatisticsOptions API

Summary:
We can't take away or change a previously available public API function.
To avoid future changes, intruduce an option class that we can extend without modifying existing APIs.

clayborg commented on lldb/include/lldb/API/SBStatisticsOptions.h:
```diff
@@ -0,0 +1,31 @@
+
+namespace lldb {
+
+class LLDB_API SBStatisticsOptions {
```
Would be good to get some header documentation on this simple class and explain which function it is for.

clayborg commented on lldb/include/lldb/Target/Statistics.h:
```diff
@@ -130,10 +130,14 @@ struct ConstStringStats {
 class TargetStats {
 public:
-  llvm::json::Value ToJSON(Target &target);
+  llvm::json::Value ToJSON(Target &target, bool summary_only = false);
```
We should still pass in a "const StatisticsOptions options = StatisticsOptions()" instead of a boolean here. If we add more options in the future, we should avoid changing this API that way.

clayborg commented on lldb/source/API/SBTarget.cpp:
```diff
@@ -198,16 +198,24 @@ SBDebugger SBTarget::GetDebugger() const {
 SBStructuredData SBTarget::GetStatistics() {
+   LLDB_INSTRUMENT_VA(this);
+  SBStatisticsOptions options;
+  options.SetSummaryOnly(false);
```
This defaults to false right? I would make sure the default `SBStatisticsOptions` constructor sets everything to good defaults and we can remove this line.

clayborg commented on lldb/include/lldb/API/SBStatisticsOptions.h:
```diff
@@ -0,0 +1,31 @@
+  const SBStatisticsOptions &operator=(const lldb::SBStatisticsOptions &rhs);
+
+  void SetSummaryOnly(bool b);
+  lldb_private::StatisticsOptions GetStatisticsOptions();
```
We don't want this function to be public, it should be protected, and not return an instance of `lldb_private::StatisticsOptions`, cause if we do this, then we need to include the header file that defines this data structure and it should be opaque. In a class like SBAddress, we do something like this:
```
class SBStatisticsOptions {
  ...
protected:
  friend class SBTarget;
  const lldb_private::StatisticsOptions &ref() const;
}
```
If this is a `lldb_private::StatisticsOptions &`, then we don't need to include the header file that defines this structure in this file since the size of a references is just a pointer size.

clayborg commented on lldb/source/API/SBStatisticsOptions.cpp:
```diff
@@ -0,0 +1,46 @@
+  m_opaque_up->summary_only = b;
+}
+
+lldb_private::StatisticsOptions SBStatisticsOptions::GetStatisticsOptions() {
+  return *m_opaque_up;
+}
```
```
const lldb_private::StatisticsOptions *SBStatisticsOptions::ref() {
  return *m_opaque_up;
}
```

clayborg commented on lldb/source/API/SBTarget.cpp:
```diff
@@ -198,16 +198,24 @@ SBDebugger SBTarget::GetDebugger() const {
-                                          target_sp.get())).str();
+      llvm::formatv(
+          "{0:2}", DebuggerStats::ReportStatistics(
+                       target_sp->GetDebugger(), target_sp.get(), options.GetStatisticsOptions()))
```
```
target_sp->GetDebugger(), target_sp.get(), options.ref()))
```

clayborg commented on lldb/source/Commands/CommandObjectStats.cpp:
```diff
@@ -83,13 +86,21 @@ class CommandObjectStatsDump : public CommandObjectParsed {
+    }
+
     bool m_all_targets = false;
+    bool m_summary_only = false;
```
Change this ivar from:
```
bool summary_only = false;
```
to:
```
lldb_private::StatisticsOptions m_stats_options;
```
And fill things into this struct when parsing the options.

clayborg commented on lldb/source/Commands/CommandObjectStats.cpp:
```diff
@@ -83,13 +86,21 @@ class CommandObjectStatsDump : public CommandObjectParsed {
     }
 
+    StatisticsOptions GetStatisticsOptions() {
+      StatisticsOptions options;
+      options.summary_only = m_summary_only;
+      return options;
```
See my ivar comment below, then this becomes:
```
return m_stats_options;
```

clayborg commented on lldb/source/Commands/CommandObjectStats.cpp:
```diff
@@ -75,6 +75,9 @@ class CommandObjectStatsDump : public CommandObjectParsed {
         m_all_targets = true;
         break;
+      case 's':
+        m_summary_only = true;
```
See comment about storing a different ivar below, then this becomes:
```
m_stats_options.summary_only = true;
```

clayborg commented on lldb/source/Commands/CommandObjectStats.cpp:
```diff
@@ -83,13 +86,21 @@ class CommandObjectStatsDump : public CommandObjectParsed {
 
     void OptionParsingStarting(ExecutionContext *execution_context) override {
       m_all_targets = false;
+      m_summary_only = false;
```
m_stats_options = StatisticsOptions();

kusmour committed:
Update test to break before dumping statistics

clayborg commented on lldb/source/Commands/CommandObjectStats.cpp:
```diff
@@ -83,13 +86,17 @@ class CommandObjectStatsDump : public CommandObjectParsed {
       return llvm::ArrayRef(g_statistics_dump_options);
     }
 
+    StatisticsOptions GetStatisticsOptions() { return m_stats_options; }
```
return a `const StatisticsOptions&` if you want to. The struct isn't big now, so no big deal.

clayborg commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -184,8 +219,12 @@ void TargetStats::IncreaseSourceMapDeduceCount() {
+    Debugger &debugger, Target *target,
+    const lldb_private::StatisticsOptions &options) {
+
+  bool summary_only = options.summary_only;
```
`const bool ...`

clayborg commented on lldb/source/Target/Statistics.cpp:
```diff
@@ -100,60 +101,94 @@ llvm::json::Value ConstStringStats::ToJSON() const {
+                    const lldb_private::StatisticsOptions &options) {
+  json::Object target_metrics_json;
+  ProcessSP process_sp = target.GetProcessSP();
+  bool summary_only = options.summary_only;
```
`const bool ...`

clayborg reviewed (approved):
A few nits you can fix if you want to.

kusmour committed:
Address nit comments
